datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java

```diff
@@ -728,6 +731,73 @@ private void removeTaskNode(String instance, String name) {
     _zkclient.deleteRecursive(instancePath);
   }
 
+  /**
+   * Update the task assignment for the given instances and issue assignment tokens for stopping tasks.
+   * Assignment tokens are used by the follower hosts to signal the leader that they finished handling their assignments.
+   *
+   * In addition to the work done by updateAllAssignments, this method writes to
+   * <ol>
+   *    *  <li>{@code /<cluster>/dms/<datastream-name>/assignmentTokens/<token1>,<token2>...}</li>
+   * </ol>
+   */
+  public void updateAllAssignmentsAndIssueTokens(Map<String, List<DatastreamTask>> assignmentsByInstance,
+      List<DatastreamGroup> stoppingDatastreamGroups) {
+    Validate.notNull(assignmentsByInstance, "null assignmentsByInstance");
+    Validate.notNull(stoppingDatastreamGroups, "null stoppingDatastreamGroups");
+
+    Map<String, Set<DatastreamTask>> currentAssignment = getAllAssignedDatastreamTasks();
+    Set<String> stoppingDatastreamTaskPrefixes = stoppingDatastreamGroups.stream().
+        map(DatastreamGroup::getTaskPrefix).collect(toSet());
+    Map<String, DatastreamGroup> taskPrefixDatastreamGroups = stoppingDatastreamGroups.stream().
+        collect(Collectors.toMap(DatastreamGroup::getTaskPrefix, Function.identity()));
+
+    // For each stopping datastream group, find all the instances that currently have tasks assigned
+    Map<DatastreamGroup, Set<String>> stoppingDgInstances = new HashMap<>();
+    for (String instance : currentAssignment.keySet()) {
```

Comment:
I think you can replace this for loop with following block, this way you do not need to create intermediate Set for instanceTasks and List for stoppingTasks and iterate on that list again.
```suggestion
    currentAssignment.keySet()
        .forEach(i -> currentAssignment.get(i).stream()
            .filter(t -> stoppingDatastreamTaskPrefixes.contains(t.getTaskPrefix()))
            .forEach(st -> {
              DatastreamGroup datastreamGroup = taskPrefixDatastreamGroups.get(st.getTaskPrefix());
              stoppingDgInstances.computeIfAbsent(datastreamGroup, k -> new HashSet<>()).add(i);
            }));
```